demo/angular-ngFor.ts

```diff
@@ -1,66 +1,122 @@
+      </div>
+    </div>
+  `,
+  styles: [
+    `
+      // !!!IMPORTANT!!! Import this through your styles.scss or angular.json! This is just for demo purposes
+      :host {
+        ::ng-deep {
+          @import "gridstack/dist/gridstack.min";
+        }
+      }
+      .grid-stack {
+        margin: 15px;
+        background-color: #d66666;
+        min-height: calc(100vh - 50px);
+      }
+
+      .grid-stack-item {
+        border: 1px solid #d1d1d1;
+      }
+    `,
+  ],
+})
+export class AngularNgForTestComponent implements AfterViewInit {
+  @ViewChildren("gridStackItem") gridstackItems: QueryList<any>;
   @Input() public items: GridStackWidget[] = [
-     { x: 0, y: 0, w: 9, h: 6},
-     { x: 9, y: 0, w: 3, h: 3},
-     { x: 9, y: 3, w: 3, h: 3},
-   ];
-   private grid: GridStack;
- 
-   constructor() {}
- 
-   // wait until after DOM is ready to init gridstack - can't be ngOnInit() as angular ngFor needs to run first!
-   public ngAfterViewInit() {
-     this.grid = GridStack.init({
-       cellHeight: 70,
-     });
-   }
- 
-   /**
-    * this would be easier with addWidget(), removeWidget(), update() but simulating angular change detection instead...
-    */
-   public add() {
-     // this SHOULD trigger ngOnChanges() but not seeing it... and doing ngDoCheck() seem extreme ?
-     // https://www.reddit.com/r/angular/comments/azjefs/change_detection_for_arraysobjects/
-     // https://angular.io/guide/lifecycle-hooks#docheck
-     this.items = [...this.items, {x: 1, y: 6, w: 3}];
-     // this.items.push({x: 1, y: 6, w: 3});
-   }
-   public delete() {
-     this.items.pop(); // todo
-   }
-   public change() {
-     this.items[0].w = 1; // todo
-   }
- 
-   public ngOnChanges(changes: SimpleChanges) {
-     if (changes.items) {
-       // TODO: ... figure what is new and call makeWidget(), old -> removeWidget(el,false) and changed -> update()
-       console.log('items changed');
-     }
-   }
- }
- 
\ No newline at end of file
+    { x: 0, y: 0, w: 1, h: 1 },
+    { x: 1, y: 1, w: 1, h: 1 },
+    { x: 2, y: 2, w: 1, h: 1 },
+  ];
+
+  private grid: GridStack;
+  private widgetToMake: Subject<{
+    action: "add" | "remove" | "update";
+    id: number;
+  }> = new Subject(); // consider to use a statemanagement like ngrx component store to do this
+
+  constructor() {}
+
+  // wait until after DOM is ready to init gridstack - can't be ngOnInit() as angular ngFor needs to run first!
+  public ngAfterViewInit() {
+    this.grid = GridStack.init({
+      alwaysShowResizeHandle:
+        /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
+          navigator.userAgent
+        ),
+      margin: 5,
+      float: true,
+    });
+
+    // To sync dom manipulation done by Angular and widget manipulation done by gridstack we need to zip the observables
+    zip(this.gridstackItems.changes, this.widgetToMake).subscribe(
+      ([changedWidget, widgetToMake]) => {

```

Comment:
With external changes, you mean like through the @Input? I didn´t focus on that. Might be a thing to improve. For that case the zip wouldn´t work. Or you have to listen to changes of the Input property by making it a setter/getter version.

Reply:
ok, then why bother with that zip thing at all then ? I mean your "public add()" method calls `makeWidget()` indirectly which my comment said was trivial to do... the thing I was trying to catch is if someone else changes the array, angular would detect those changes but gridstack would not. I can just update my version to do what you are in a much simpler case, and maybe add a setter as well to do a load() which diff the changes already.

Reply:
We need the zip thing because else you don´t know when Angular has finished its dom manipulation. Without the zip you would have to use setTimeout for a tiny bit of a second to wait for Angular and then call makeWidget which is an unsolid hack.